api/resources_portal/views/document_views.py

```diff
@@ -0,0 +1,316 @@
+        I used the highest possible precision threshold, but this might increase the amount
+        of memory used.
+        """
+        facets = self.construct_facets(request, view)
+        for field, facet in iteritems(facets):
+            agg = facet["facet"].get_aggregation()
+            queryset.aggs.bucket(field, agg).metric(
+                "total_samples", "cardinality", field="id", precision_threshold=40000,
+            )
+        return queryset
+
+
+##
+# ElasticSearch powered Search and Filter
+##
+@method_decorator(
+    name="list",
+    decorator=swagger_auto_schema(
+        manual_parameters=[
+            openapi.Parameter(
+                name="category",
+                in_=openapi.IN_QUERY,
+                type=openapi.TYPE_STRING,
+                description="Allows filtering the results by category, can have multiple values. Eg: `?category=plasmid&category=other`",
+            ),
+            openapi.Parameter(
+                name="organism",
+                in_=openapi.IN_QUERY,
+                type=openapi.TYPE_STRING,
+                description="Allows filtering the results by organism",
+            ),
+            openapi.Parameter(
+                name="has_publication",
+                in_=openapi.IN_QUERY,
+                type=openapi.TYPE_STRING,
+                description="Allows filtering the results by has_publication",
+            ),
+            openapi.Parameter(
+                name="has_pre_print",
+                in_=openapi.IN_QUERY,
+                type=openapi.TYPE_STRING,
+                description="Allows filtering the results by has_pre_print",
+            ),
+        ],
+        operation_description="""
+Use this endpoint to search among the experiments.
+
+This is powered by ElasticSearch, information regarding advanced usages of the
+filters can be found in the [Django-ES-DSL-DRF docs](https://django-elasticsearch-dsl-drf.readthedocs.io/en/0.17.1/filtering_usage_examples.html#filtering)
+
+There's an additional field in the response named `facets` that contain stats on the number of results per filter type.
+
+Example Requests:
+```
+?search=zebrafish
+?id=1
+?search=zebrafish&category=plasmid&has_publication=true
+?ordering=source_first_published
+```
+""",
+    ),
+)
+
+# Defines search and filter fields for Materials
+class MaterialDocumentView(DocumentViewSet):
+    document = MaterialDocument
+    serializer_class = MaterialDocumentSerializer
+    pagination_class = ESLimitOffsetPagination
+
+    # Filter backends provide different functionality we want
+    filter_backends = [
+        FilteringFilterBackend,
+        OrderingFilterBackend,
+        DefaultOrderingFilterBackend,
+        CompoundSearchFilterBackend,
+        FacetedSearchFilterBackendExtended,
+    ]
+    lookup_field = "id"
+
+    # Define search fields
+    # Is this exhaustive enough?
```

Comment:
I think so!

Reply:
Sweet then I don't think we need the comment any more!

Reply:
removed.